File: roundTrip2.cpp
#include<bits/stdc++.h>
using namespace std;
#define fast_io ios_base::sync_with_stdio(false);cin.tie(NULL)

void init() {
fast_io;
#ifndef ONLINE_JUDGE
freopen("input.txt", "r", stdin);
freopen("output.txt", "w", stdout);
#endif // ONLINE_JUDGE
}


bool isCycle(int node, vector<int>&vis, vector<int>&recStack, stack<int>&st, vector<int>adj[]){

	     vis[node] = recStack[node] = 1;

	     st.push(node);

	     for(auto neighbor: adj[node]){
	     	   if(!vis[neighbor]){ // include it into the cycle
	     	   	    if(isCycle(neighbor, vis, recStack, st, adj)){
	     	   	    	  return true;
	     	   	    }
	     	   }else if(recStack[neighbor]){
	     	   	    // back-edge
	     	   	    st.push(neighbor);
	     	   	    return true;
	     	   }
	     }

	     recStack[node]=0;
	     st.pop();

	     return false;
}

vector<int>extract(stack<int>&st){

	   int node = st.top();

	   vector<int>cycle;

	   cycle.push_back(node);

	   st.pop();

	   while(!st.empty()){
	   	   cycle.push_back(st.top());
           if(st.top() == node){
           	  break;
           }
           st.pop();
	   }

	   return cycle;
}
 
//just detect if there is a cycle or not 
vector<int>roundTrip(int n, vector<int>adj[]){
      
      vector<int>vis(n+1,0);
      vector<int>recStack(n+1,0);
      stack<int>st;

      for(int i=1;i<=n;i++){
      	   if(!vis[i]){ 
                 if(isCycle(i, vis, recStack, st, adj)){
                 	  return extract(st);
                 }
      	   }
      }

      return {};
}

int main(){
   init();
   
   int n,m;
   cin >> n >> m;

   vector<int>adj[n+1];

   for(int i=0;i<m;i++){
   	  int a,b;
   	  cin >> a >> b;
   	  adj[a].push_back(b);
   }

   vector<int>cycle = roundTrip(n, adj);

   if(cycle.size() == 0){
   	  cout << "IMPOSSIBLE\n";
   	  return 0;
   }
  
   reverse(cycle.begin(), cycle.end());
   cout << cycle.size() << "\n";

   for(auto it: cycle)
   	  cout << it << " ";
    
}

File: roundTrip2.py
import sys

sys.setrecursionlimit(10**6)

def init():
    pass

def isCycle(node, vis, recStack, st, adj):
    vis[node] = recStack[node] = 1
    st.append(node)

    for neighbor in adj[node]:
        if vis[neighbor] == 0:  # include it into the cycle
            if isCycle(neighbor, vis, recStack, st, adj):
                return True
        elif recStack[neighbor]:
            # back-edge
            st.append(neighbor)
            return True

    recStack[node] = 0
    st.pop()

    return False

def extract(st):
    node = st[-1]
    cycle = [node]
    st.pop()

    while st:
        cycle.append(st[-1])
        if st[-1] == node:
            break
        st.pop()

    return cycle

def roundTrip(n, adj):
    vis = [0] * (n + 1)
    recStack = [0] * (n + 1)
    st = []

    for i in range(1, n + 1):
        if not vis[i]:
            if isCycle(i, vis, recStack, st, adj):
                return extract(st)

    return []

def main():
    init()

    n, m = map(int, input().split())

    adj = [[] for _ in range(n + 1)]

    for _ in range(m):
        a, b = map(int, input().split())
        adj[a].append(b)

    cycle = roundTrip(n, adj)

    if len(cycle) == 0:
        print("IMPOSSIBLE")
        return

    cycle.reverse()
    print(len(cycle))

    for it in cycle:
        print(it, end=" ")

if __name__ == "__main__":
    main()
